### ISSUE TITLE
Reducing RAM usage

### ISSUE THREAD

**Issue by PerplexedPeach** [cwtools/cwtools-vscode#91]
This extension is taking up around 4-5GB of RAM, making it hard to use together with the game open. I suspect it's because my workspace has too many CK3 projects in it, but I'm not sure if this contributes to it. Does having more CK3 files in a workspace (without opening) them increase RAM usage? My workspace includes:

- base game's game directory
- my mods
- mods that I have compatibility patches for
- mods that my mods depend on

Another possible culprit is that many of my event files are autogenerated via a python script, so I have like 20k line files that should be excluded from processing. Is there a way to mark files or directories as excluded from indexing?

**Comment by tboby** [comment]
> Does having more CK3 files in a workspace (without opening) them increase RAM usage? 

Yes: cwtools keeps a processed form of every script and localisation file in it's scope in memory. 

It's basically unavoidable to have anything responsive, and all programming languages have high memory usage if you want to open several projects at once sadly!

At a minimum this would be the "vanilla cache" which contains all the vanilla files, plus your mod. If you're including the base game in your workspace you might be causing it to load all of vanilla again. If your mod is a total conversion, you'd then effectively be keeping three copies of vanilla around :)

One recommendation I have would be to create multiple workspaces, where some contain only your core mods + dependencies. This is the approach taken in other languages when solutions get too large.

The huge files will cause some problems, although if they're over 2MB they're probably already being ignored by the default max file size setting `cwtools.maxFileSize`. There's also `cwtools.ignore_patterns` which you could use to ignore whole filepaths, which you should be able to use to ignore your reference inclusion of vanilla.

**Comment by PerplexedPeach** [comment]
I see, thanks! Those are all actionable and I'll get back to you with how well some of them work.

**Comment by PerplexedPeach** [comment]
Removing the base game and some other mods from the workspace reduced it down to around 3.5GB of RAM, so that worked as intended. The `cwtools.ignore_patterns` doesn't seem to be working. When I put a single string there (not even utilizing any glob patterns; it's just the path relative to the workspace of one of my event files), the extension just stops working - the cwtools loaded files tile in the bottom left is completely empty and it makes no attempt to load anything. Is there a way to debug the extension settings?
